### ISSUE TITLE
Metrics Improvement Project

### ISSUE THREAD

**Issue by dannyl1u** [apache/airflow#42881]
### Description

From @ferruzzi:

> Currently when you add a new metric to the codebase, you must also manually update the docs page.  The docs page inevitably gets out of date and misses some details.  We want an automated system to generate the docs page based on the actual metrics.  There are also known instances where the same metric is being created and emitted in more than one place, causing duplicate data.  These will have to be fixed manually and an automated check might possibly (stretch goal?)  include checking for same or ”too similar” names while collecting the names for the docs page.

> Phase 1
> Situation:
> We support multiple different Metrics backends [0].  The two main ones are StatsD and OpenTelemetry.  This is managed though an interface class [1] which is implemented for each backend (examples:  StatsD[2] and OTel[3]).   StatsD was the only supported version well into Airflow 2.x and the entire codebase was designed with StatsD in mind so it was a good chunk of work to abstract it out and there are a few remaining tasks to perfect the new implementation.
> Task 1:
> StatsD has a name length limit of around 300 characters.  OTel limits names to 34 characters, but allows tagging.  Our temporary solution was to emit almost everything twice, once in the long format for StatsD and again in the short format with tags for OTel.  We also had to add code [4] to make sure the name is safe for OTel, and other hacks to make it work.
> The first task in this project is to understand the difference in how the two implementations handle their names and them add a "get_name" method to the interface: `def get_name(metric_name: str, tags: dict[str: str])`.  In the statsd_logger [2] implementation it will concatenate the tags onto the name and in the OTel implementation it will just return name.
> Once that is implemented, it can be used in the various emit methods (incr, decr, etc) instead of all the name validation code, and search the code for places where we are emitting things more than once and clean it up.
> Example:
> You can see an example in local_task_job_runner [5].  We emit `local_task_job.task_exit.<job_id>.<dag_id>.<task_id>.<return_code>` for StatsD but that results in a name too long for OTel so we also emit `local_task_job.task_exit`, and the name validation method [4] in the OTel implementation catches the one that is too long and just swallows it.  What we should do instead is pass incr() the name and the tags and let StatsD and OTel handle them accordingly.
> [0] https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/metrics.html#metric-descriptions
> [1] https://github.com/apache/airflow/blob/main/airflow/metrics/base_stats_logger.py
> [2] https://github.com/apache/airflow/blob/main/airflow/metrics/statsd_logger.py
> [3] https://github.com/apache/airflow/blob/main/airflow/metrics/otel_logger.py
> [4] https://github.com/apache/airflow/blob/main/airflow/metrics/otel_logger.py#L128
> [5] https://github.com/apache/airflow/blob/main/airflow/jobs/local_task_job_runner.py#L352

### Use case/motivation

From @ferruzzi:

> Currently when you add a new metric to the codebase, you must also manually update the docs page.  The docs page inevitably gets out of date and misses some details.  We want an automated system to generate the docs page based on the actual metrics.  There are also known instances where the same metric is being created and emitted in more than one place, causing duplicate data.  These will have to be fixed manually and an automated check might possibly (stretch goal?)  include checking for same or ”too similar” names while collecting the names for the docs page.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by ashb** [comment]
This sounds great. Could you sketch out the sort of API you are thinking of, and give an example use from somewhere in scheduler job please?

**Comment by dannyl1u** [comment]
Hi @ashb, could you clarify what you mean by the use in the scheduler job? 

The first step is refactoring the metrics code so that https://github.com/apache/airflow/blob/main/airflow/metrics/base_stats_logger.py will have a `get_name` method that https://github.com/apache/airflow/blob/main/airflow/metrics/datadog_logger.py, https://github.com/apache/airflow/blob/main/airflow/metrics/otel_logger.py, https://github.com/apache/airflow/blob/main/airflow/metrics/statsd_logger.py, etc. will inherit the method for their own specific naming conventions. This will improve the code by not needing specific name validators for each implementation https://github.com/apache/airflow/blob/main/airflow/metrics/otel_logger.py#L128.

cc'ing @ferruzzi in case I'm missing something

Then, once the above is done, we will plan the next step to build an automated system to generate docs based on the actual metrics.

cc @arshiazr

**Comment by potiuk** [comment]
@ferruzzi @howardyoo -> I think you two should be quite a bit involved in the review and ideas here. 

Initially when we implemented open-telemetry metrics  we thought we could do it in the way that we could emit legacy 'statsd" metrics using opentelemetry interface (basically implement or use some kind of `opentelemetry -> statsd` bridge), because of some implementation details it turned out to be impossible (or difficult).

However that can still be explored, maybe  that is still a possiblity? Implementing our own interface that wraps both Opentelemetry Metrics and Statsd one is of course a possibility, but (at least intuitively - without knowing all the details) - it could be that opentelemetry API could be used to emit the legacy statsd events in "mostly" compatible way. 

And implementing it in Airflow 3 gives us also an opportunity for the "mostly" part. If we can get 90% of the backwards-compatible statsd interface in place and only "few", "less important" metrics changed to be incompatible, maybe that is a way to go?

**Comment by ashb** [comment]
Scheduler job = https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job_runner.py

**Comment by ArshiaZr** [comment]
I have just raised a new PR  (https://github.com/apache/airflow/pull/43018). I'd be grateful if you can review it.

cc @ferruzzi

**Comment by potiuk** [comment]
@dannyl1u We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by amoghrajesh** [comment]
Since this has not yet been picked up and we are approaching the 3.2 release, I am moving it to the 3.3 milestone.
